core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -788,7 +804,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
 
       // triggering another task for compaction task only if it succeeds
       if (succeeded) {
-        submitLogProcessTask(rootPath) { () => compact(reader) }
+        submitLogProcessTask(rootPath, compactExecutor) { () => compact(reader) }
```

Comment:
It's actually a MUST change. Otherwise, submitting a task to the self pool will cause a deadlock if the queue is full.